### PR TITLE
Enable VPC flow logs (#4176)

### LINKED ISSUE #4176: Enable VPC flow logs
Data Browser System Overview - Logging   Monitoring (15)

https://docs.aws.amazon.com/vpc/latest/userguide/flow-logs-cwl.html#flow-logs-cwl-create-flow-log

https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/flow_log

Enable for all traffic, watch log volume, adjust settings if needed.

### PR DESCRIPTION
**Pull request by achave11-ucsc**
Connected issues: #4176 

### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references all connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] For each connected issue, there is at least one commit whose title references that issue
- [x] PR is connected to all connected issues via ZenHub
- [x] PR description links to connected issues
- [x] Added `partial` label to PR <sub>or this PR completely resolves all connected issues</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title

### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR <sub>or this PR does not require reindexing</sub>
- [x] Added `a` (compatible changes) or `A` (incompatible ones) tag to commit title <sub>or this PR does not modify the Azul service API</sub>
- [x] Added `API` label to connected issues <sub>or this PR does not modify the Azul service API</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR <sub>or this PR does not touch requirements*.txt</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>

### Peer reviewer (after requesting changes)

Uncheck the *Author (before every review)* checklists.

### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *before every review* checklists. Update the `N reviews` label.

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Pushed PR branch to GitLab `dev` and added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved connected issues to Merged column in ZenHub
- [x] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [x] Shortened the PR chain <sub>or this PR is not labeled `base`</sub>
- [x] Deploy to the `dev.gitlab` componet.
- [x] Deploy to the `dev.shared` componet.
- [x] Pushed merge commit to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Deploy to the `anvildev.gitlab` componet.
- [x] Deploy to the `anvildev.shared` componet.
- [x] Pushed merge commit to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is*
pushed determines this checklist item.

### Operator (reindex)

- [x] Deleted unreferenced indices in `dev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvildev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `dev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvildev` deployment <sub>or this PR does not require reindexing</sub>

### Operator

- [x] Add checklist items to promotion PR; Deploy the `prod.gitlab` followed by the `prod.shared` component
- [x] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Comment by achave11-ucsc** [comment]
The following is the TF plan output for the newly added resources.
```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated
with the following symbols:
  + create

Terraform will perform the following actions:

  # aws_cloudwatch_log_group.gitlab_vpc will be created
  + resource "aws_cloudwatch_log_group" "gitlab_vpc" {
      + arn               = (known after apply)
      + id                = (known after apply)
      + name              = "/aws/vpc/azul-gitlab"
      + retention_in_days = 180
      + tags              = {
          + "Name"                = "azul-gitlab_vpc"
          + "billing"             = "hca"
          + "component"           = "azul-gitlab_vpc"
          + "deployment"          = "dev"
          + "owner"               = "hannes@ucsc.edu"
          + "service"             = "azul"
          + "terraform_component" = "gitlab"
        }
      + tags_all          = {
          + "Name"                = "azul-gitlab_vpc"
          + "billing"             = "hca"
          + "component"           = "azul-gitlab_vpc"
          + "deployment"          = "dev"
          + "owner"               = "hannes@ucsc.edu"
          + "service"             = "azul"
          + "terraform_component" = "gitlab"
        }
    }

  # aws_flow_log.gitlab will be created
  + resource "aws_flow_log" "gitlab" {
      + arn                      = (known after apply)
      + iam_role_arn             = (known after apply)
      + id                       = (known after apply)
      + log_destination          = (known after apply)
      + log_destination_type     = "cloud-watch-logs"
      + log_format               = (known after apply)
      + log_group_name           = (known after apply)
      + max_aggregation_interval = 600
      + tags                     = {
          + "Name"                = "azul-gitlab"
          + "billing"             = "hca"
          + "component"           = "azul-gitlab"
          + "deployment"          = "dev"
          + "owner"               = "hannes@ucsc.edu"
          + "service"             = "azul"
          + "terraform_component" = "gitlab"
        }
      + tags_all                 = {
          + "Name"                = "azul-gitlab"
          + "billing"             = "hca"
          + "component"           = "azul-gitlab"
          + "deployment"          = "dev"
          + "owner"               = "hannes@ucsc.edu"
          + "service"             = "azul"
          + "terraform_component" = "gitlab"
        }
      + traffic_type             = "ALL"
      + vpc_id                   = "vpc-03aeb4a3542c7a5d3"
    }

  # aws_iam_policy.gitlab_vpc will be created
  + resource "aws_iam_policy" "gitlab_vpc" {
      + arn       = (known after apply)
      + id        = (known after apply)
      + name      = "azul-gitlab_vpc"
      + path      = "/"
      + policy    = jsonencode(
            {
              + Statement = [
                  + {
                      + Action   = [
                          + "logs:CreateLogGroup",
                          + "logs:CreateLogStream",
                          + "logs:PutLogEvents",
                          + "logs:DescribeLogGroups",
                          + "logs:DescribeLogStreams",
                        ]
                      + Effect   = "Allow"
                      + Resource = "*"
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + policy_id = (known after apply)
      + tags      = {
          + "Name"                = "azul-gitlab_vpc"
          + "billing"             = "hca"
          + "component"           = "azul-gitlab_vpc"
          + "deployment"          = "dev"
          + "owner"               = "hannes@ucsc.edu"
          + "service"             = "azul"
          + "terraform_component" = "gitlab"
        }
      + tags_all  = {
          + "Name"                = "azul-gitlab_vpc"
          + "billing"             = "hca"
          + "component"           = "azul-gitlab_vpc"
          + "deployment"          = "dev"
          + "owner"               = "hannes@ucsc.edu"
          + "service"             = "azul"
          + "terraform_component" = "gitlab"
        }
    }

  # aws_iam_role.gitlab_vpc will be created
  + resource "aws_iam_role" "gitlab_vpc" {
      + arn                   = (known after apply)
      + assume_role_policy    = jsonencode(
            {
              + Statement = [
                  + {
                      + Action    = "sts:AssumeRole"
                      + Effect    = "Allow"
                      + Principal = {
                          + Service = "vpc-flow-logs.amazonaws.com"
                        }
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + create_date           = (known after apply)
      + force_detach_policies = false
      + id                    = (known after apply)
      + managed_policy_arns   = (known after apply)
      + max_session_duration  = 3600
      + name                  = "azul-gitlab_vpc"
      + name_prefix           = (known after apply)
      + path                  = "/"
      + tags                  = {
          + "Name"                = "azul-gitlab_vpc"
          + "billing"             = "hca"
          + "component"           = "azul-gitlab_vpc"
          + "deployment"          = "dev"
          + "owner"               = "hannes@ucsc.edu"
          + "service"             = "azul"
          + "terraform_component" = "gitlab"
        }
      + tags_all              = {
          + "Name"                = "azul-gitlab_vpc"
          + "billing"             = "hca"
          + "component"           = "azul-gitlab_vpc"
          + "deployment"          = "dev"
          + "owner"               = "hannes@ucsc.edu"
          + "service"             = "azul"
          + "terraform_component" = "gitlab"
        }
      + unique_id             = (known after apply)

      + inline_policy {
          + name   = (known after apply)
          + policy = (known after apply)
        }
    }

  # aws_iam_role_policy_attachment.gitlab_vpc will be created
  + resource "aws_iam_role_policy_attachment" "gitlab_vpc" {
      + id         = (known after apply)
      + policy_arn = (known after apply)
      + role       = "azul-gitlab_vpc"
    }

Plan: 5 to add, 0 to change, 0 to destroy.

──────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Note: You didn't use the -out option to save this plan, so Terraform can't guarantee to take exactly these actions
if you run "terraform apply" now.

```

File: terraform/gitlab/gitlab.tf.json.template.py [hunk 1]
```diff
@@ -897,6 +900,19 @@ def qq(*words):
...
                 'name': '/aws/vpn/azul-gitlab',
                 'retention_in_days': config.audit_log_retention_days,
+            },
+            'gitlab_vpc': {
+                'name': '/aws/vpc/azul-gitlab',
+                'retention_in_days': config.audit_log_retention_days,
+            }
+        },
+        'aws_flow_log': {
+            'gitlab_vpc': {
```
Comment (achave11-ucsc) [inline terraform/gitlab/gitlab.tf.json.template.py]:
```suggestion
            'gitlab': {
```

File: terraform/gitlab/gitlab.tf.json.template.py [hunk 2]
```diff
@@ -1161,7 +1192,26 @@ def qq(*words):
                 'name': config.permissions_boundary_name,
                 'path': '/',
                 'policy': '${data.aws_iam_policy_document.gitlab_boundary.json}'
-            }
+            },
+            'gitlab_vpc': {
+                'name': 'azul-gitlab-vpc',
```
Comment (achave11-ucsc) [inline terraform/gitlab/gitlab.tf.json.template.py]:
```suggestion
                'name': 'azul-gitlab_vpc',
```

**Comment by achave11-ucsc** [comment]
Enabling flow logs for the default VPC in the `shared` component produced a valid TF plan.
```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following
symbols:
  + create

Terraform will perform the following actions:

  # aws_cloudwatch_log_group.default will be created
  + resource "aws_cloudwatch_log_group" "default" {
      + arn               = (known after apply)
      + id                = (known after apply)
      + name              = "/aws/vpc/azul-default-dev"
      + retention_in_days = 180
      + tags              = {
          + "Name"                = "azul-default"
          + "billing"             = "hca"
          + "component"           = "azul-default"
          + "deployment"          = "dev"
          + "owner"               = "hannes@ucsc.edu"
          + "service"             = "azul"
          + "terraform_component" = "shared"
        }
      + tags_all          = {
          + "Name"                = "azul-default"
          + "billing"             = "hca"
          + "component"           = "azul-default"
          + "deployment"          = "dev"
          + "owner"               = "hannes@ucsc.edu"
          + "service"             = "azul"
          + "terraform_component" = "shared"
        }
    }

  # aws_flow_log.default will be created
  + resource "aws_flow_log" "default" {
      + arn                      = (known after apply)
      + iam_role_arn             = (known after apply)
      + id                       = (known after apply)
      + log_destination          = (known after apply)
      + log_destination_type     = "cloud-watch-logs"
      + log_format               = (known after apply)
      + log_group_name           = (known after apply)
      + max_aggregation_interval = 600
      + tags                     = {
          + "Name"                = "azul-default"
          + "billing"             = "hca"
          + "component"           = "azul-default"
          + "deployment"          = "dev"
          + "owner"               = "hannes@ucsc.edu"
          + "service"             = "azul"
          + "terraform_component" = "shared"
        }
      + tags_all                 = {
          + "Name"                = "azul-default"
          + "billing"             = "hca"
          + "component"           = "azul-default"
          + "deployment"          = "dev"
          + "owner"               = "hannes@ucsc.edu"
          + "service"             = "azul"
          + "terraform_component" = "shared"
        }
      + traffic_type             = "ALL"
      + vpc_id                   = "vpc-aadf0dd0"
    }

  # aws_iam_role.default will be created
  + resource "aws_iam_role" "default" {
      + arn                   = (known after apply)
      + assume_role_policy    = jsonencode(
            {
              + Statement = [
                  + {
                      + Action    = "sts:AssumeRole"
                      + Effect    = "Allow"
                      + Principal = {
                          + Service = "vpc-flow-logs.amazonaws.com"
                        }
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + create_date           = (known after apply)
      + force_detach_policies = false
      + id                    = (known after apply)
      + managed_policy_arns   = (known after apply)
      + max_session_duration  = 3600
      + name                  = "azul-default_vpc-dev"
      + name_prefix           = (known after apply)
      + path                  = "/"
      + tags                  = {
          + "Name"                = "azul-default"
          + "billing"             = "hca"
          + "component"           = "azul-default"
          + "deployment"          = "dev"
          + "owner"               = "hannes@ucsc.edu"
          + "service"             = "azul"
          + "terraform_component" = "shared"
        }
      + tags_all              = {
          + "Name"                = "azul-default"
          + "billing"             = "hca"
          + "component"           = "azul-default"
          + "deployment"          = "dev"
          + "owner"               = "hannes@ucsc.edu"
          + "service"             = "azul"
          + "terraform_component" = "shared"
        }
      + unique_id             = (known after apply)

      + inline_policy {
          + name   = (known after apply)
          + policy = (known after apply)
        }
    }

  # aws_iam_role_policy.default will be created
  + resource "aws_iam_role_policy" "default" {
      + id     = (known after apply)
      + name   = "azul-default_vpc-dev"
      + policy = jsonencode(
            {
              + Statement = [
                  + {
                      + Action   = [
                          + "logs:CreateLogGroup",
                          + "logs:CreateLogStream",
                          + "logs:PutLogEvents",
                          + "logs:DescribeLogGroups",
                          + "logs:DescribeLogStreams",
                        ]
                      + Effect   = "Allow"
                      + Resource = "*"
                    },
                ]
              + Version   = "2012-10-17"
            }
        )
      + role   = (known after apply)
    }

Plan: 4 to add, 0 to change, 0 to destroy.

──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

Note: You didn't use the -out option to save this plan, so Terraform can't guarantee to take exactly these actions if you run
"terraform apply" now.


```

File: UPGRADING.rst [hunk 3]
```diff
@@ -29,6 +29,17 @@ Manually deploy the ``shared`` component of any main deployment just before
 pushing the merge commit to the GitLab instance in that deployment.
 
 
+#4176 Enable VPC flow logs
+==========================
+
+Operator
+~~~~~~~~
+
+Manually deploy the ``gitlab`` as well as the ``shared`` component of any main
```
Comment (nadove-ucsc) [inline UPGRADING.rst]:
```suggestion
Manually deploy the ``gitlab`` and  ``shared`` components of any main
```

File: terraform/shared/shared.tf.json.template.py [hunk 4]
```diff
@@ -105,12 +105,29 @@ def metric_name(self) -> str:
...
+        'aws_vpc': {
+            vpc.default_vpc_name: {
+                'default': True
+            }
         }
     },
     'resource': {
+        # https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/default_vpc#attributes-reference
+        # While the `aws_default_vpc` resource doesn't list `.id` as an
+        # attribute in the docs, its usage is valid.
```
Comment (nadove-ucsc) [inline terraform/shared/shared.tf.json.template.py]: I would move this comment to where the `.id` attribute is accessed

**Review by nadove-ucsc — CHANGES_REQUESTED** [review]
Just a few small things

**Review by nadove-ucsc — DISMISSED** [review]

**Review by hannes-ucsc — DISMISSED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**